source/array-element.d.ts

```diff
@@ -0,0 +1,28 @@
+/**
+Extracts the element type of an array or union of arrays.
+
+Returns `never` if T is not an array.
+
+It creates a type-safe way to access the element type of `unknown` type.
+
+@example
+```
+import type {ArrayElement} from 'type-fest';
+
+declare const getMostCommonElement: <T>(array: T[]) => ArrayElement<typeof array>;
+
+getMostCommonElement([1, 2, 3]);
+//=> 1 | 2 | 3
+
+getMostCommonElement(['foo', 'bar', 'baz'] as const);
+//=> 'foo' | 'bar' | 'baz'
+```
+
+@see {@link ArrayValues} for when you know that the input is an array.
+
+@category Array
+*/
+export type ArrayElement<ArrayType extends readonly unknown[]> =
+	ArrayType extends ReadonlyArray<infer ElementType>
+		? ElementType
+		: never;
```

Comment:
I'm not sure this works @strongpauly

> Should this be more?:
> 
> ```ts
> type ArrayElement<ArrayType extends readonly unknown[]> = 
>   ArrayType extends readonly (infer ElementType)[] ? ElementType : never;
> ```
> 
> Copy paste from [stackoverflow.com/questions/41253310/typescript-retrieve-element-type-information-from-array-type](https://stackoverflow.com/questions/41253310/typescript-retrieve-element-type-information-from-array-type)

// https://github.com/sindresorhus/type-fest/pull/906#discussion_r1673712673

Reply:
Well, as long as it returns a union of the types of the array elements, its fit for purpose, IMO.